ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryLimitFunctionalTest.java

```diff
@@ -0,0 +1,491 @@
+            .withProperties(COMMON_CONFIG)
+            .build();
+
+    @Rule
+    public final TestKsqlRestApp REST_APP_1 = TestKsqlRestApp
+            .builder(TEST_HARNESS::kafkaBootstrapServers)
+            .withProperty(KSQL_STREAMS_PREFIX + StreamsConfig.STATE_DIR_CONFIG, getNewStateDir())
+            .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:0")
+            .withProperty(KsqlRestConfig.INTERNAL_LISTENER_CONFIG, "http://localhost:" + INT_PORT_1)
+            .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:" + INT_PORT_1)
+            .withFaultyKsqlClient(APP_SHUTOFFS_1::getKsqlOutgoing)
+            .withProperty(KSQL_STREAMS_PREFIX + CONSUMER_PREFIX
+                    + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, FaultyKafkaConsumer.FaultyKafkaConsumer1.class.getName())
+            .withProperties(COMMON_CONFIG)
+            .build();
+
+    @Rule
+    public final TestKsqlRestApp REST_APP_2 = TestKsqlRestApp
+            .builder(TEST_HARNESS::kafkaBootstrapServers)
+            .withProperty(KSQL_STREAMS_PREFIX + StreamsConfig.STATE_DIR_CONFIG, getNewStateDir())
+            .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:0")
+            .withProperty(KsqlRestConfig.INTERNAL_LISTENER_CONFIG, "http://localhost:" + INT_PORT_2)
+            .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:" + INT_PORT_2)
+            .withFaultyKsqlClient(APP_SHUTOFFS_2::getKsqlOutgoing)
+            .withProperty(KSQL_STREAMS_PREFIX + CONSUMER_PREFIX
+                    + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, FaultyKafkaConsumer.FaultyKafkaConsumer2.class.getName())
+            .withProperties(COMMON_CONFIG)
+            .build();
+
+    public final TestApp TEST_APP_0 = new TestApp(HOST0, REST_APP_0, APP_SHUTOFFS_0);
+    public final TestApp TEST_APP_1 = new TestApp(HOST1, REST_APP_1, APP_SHUTOFFS_1);
+    public final TestApp TEST_APP_2 = new TestApp(HOST2, REST_APP_2, APP_SHUTOFFS_2);
+
+    @ClassRule
+    public static final RuleChain CHAIN = RuleChain
+            .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
+            .around(TEST_HARNESS)
+            .around(JAAS_CONFIG)
+            .around(TMP);
+
+    @Rule
+    public final Timeout timeout = Timeout.builder()
+            .withTimeout(4, TimeUnit.MINUTES)
+            .withLookingForStuckThread(true)
+            .build();
+
+    @BeforeClass
+    public static void setUpClass() {
+        FaultyKafkaConsumer.FaultyKafkaConsumer0.setPauseOffset(APP_SHUTOFFS_0::getKafkaPauseOffset);
+        FaultyKafkaConsumer.FaultyKafkaConsumer1.setPauseOffset(APP_SHUTOFFS_1::getKafkaPauseOffset);
+        FaultyKafkaConsumer.FaultyKafkaConsumer2.setPauseOffset(APP_SHUTOFFS_2::getKafkaPauseOffset);
+    }
+
+    @Before
+    public void setUp() {
+        //Create topic with 4 partition to control who is active and standby
```

Comment:
Heh, the cluster formation logic works because we have a single partition and it's well defined who's active and who's standby.  With 4 partitions, you're going to have each node as active for something.